src/server/pps/server/api_server.go

```diff
@@ -983,42 +997,89 @@ func (a *apiServer) GetLogs(request *pps.GetLogsRequest, apiGetLogsServer pps.AP
 					// Log message passes all filters -- return it
 					select {
 					case logChs[i] <- msg:
-					case <-done:
+					case <-ctx.Done():
 						return nil
 					}
 				}
 				return nil
-			}, backoff.New10sBackOff())
+			}, backoff.New10sBackOff(), func(err error, t time.Duration) error {
+				logrus.Errorf("error retreiving logs; retrying in %s: %v", t, err)
+				return nil
+			})
 
 			// Used up all retries -- no logs from worker
 			if err != nil {
 				select {
 				case errCh <- err:
-				case <-done:
 				default:
 				}
 			}
 		}()
 	}
 
-nextLogCh:
-	for _, logCh := range logChs {
+	// If we are not following logs, we want to print logs from the pods one
+	// by one, so that the logs appear continuous within a pod.
+	if request.Follow {
+		// Otherwise, we stream logs from all pods concurrently.
+		logCh := forwardLogs(ctx, logChs)
 		for {
-			msg, ok := <-logCh
-			if !ok {
-				continue nextLogCh
-			}
-			if err := apiGetLogsServer.Send(msg); err != nil {
+			select {
+			case msg := <-logCh:
+				if err := apiGetLogsServer.Send(msg); err != nil {
+					return err
+				}
+			case err := <-errCh:
 				return err
+			case <-ctx.Done():
+				return nil
 			}
 		}
+	} else {
+	nextLogCh:
+		for _, logCh := range logChs {
+			for {
+				msg, ok := <-logCh
+				if !ok {
+					continue nextLogCh
+				}
+				if err := apiGetLogsServer.Send(msg); err != nil {
+					return err
+				}
+			}
+		}
+
+		select {
+		case err := <-errCh:
+			return err
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		return nil
 	}
-	select {
-	case err := <-errCh:
-		return err
-	default:
+}
+
+// forwardLogs is a helper functions that aggregates logs from a number of
+// channels of logs.
+func forwardLogs(ctx context.Context, logChs []chan *pps.LogMessage) chan *pps.LogMessage {
```

Comment:
Is this function really necessary? It seems like any situation where you need to results of multiple channels to be aggregated into a single channel could be greatly simplified by just giving everyone the same channel to begin with.